# Processo Seletivo

### analise_python.ipynb - Vagner Przibiciem

In [1]:
import basedosdados as bd
import pandas as pd
from datetime import timedelta, datetime

#Exibir todas as linhas e colunas
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Importando todos os dados por meio da biblioteca basedosdados:

In [2]:
chamados = bd.read_sql("SELECT * FROM datario.administracao_servicos_publicos.chamado_1746 WHERE data_inicio >= '2023-04-01' AND data_inicio < '2023-04-02'", billing_project_id='processo-seletivo-414621')

Downloading: 100%|██████████████████████████████████████████████████████████████████| 73/73 [00:00<00:00, 249.12rows/s]


In [3]:
chamados

id_chamado         data_inicio            data_fim id_bairro  \
0    18516246 2023-04-01 00:55:38 2023-04-01 00:55:38      None   
1    18516254 2023-04-01 01:08:04 2023-04-01 01:43:38         1   
2    18516194 2023-04-01 00:04:59                 NaT        14   
3    18516238 2023-04-01 00:38:22                 NaT         5   
4    18516208 2023-04-01 00:14:00                 NaT        14   
5    18516200 2023-04-01 00:08:57                 NaT        14   
6    18516244 2023-04-01 00:51:41                 NaT       161   
7    18516186 2023-04-01 00:03:51                 NaT       161   
8    18516204 2023-04-01 00:10:15                 NaT        33   
9    18516221 2023-04-01 00:20:36                 NaT        33   
10   18516247 2023-04-01 00:57:30                 NaT        15   
11   18516181 2023-04-01 00:00:37                 NaT        26   
12   18516213 2023-04-01 00:16:10                 NaT        26   
13   18516199 2023-04-01 00:08:22                 NaT        26   
14   18516218 2023-04-01 00:18:32                 NaT        17   
15   18516261 2023-04-01 01:23:20                 NaT        26   
16   18516212 2023-04-01 00:15:18                 NaT        26   
17   18516237 2023-04-01 00:38:21                 NaT        18   
18   18516196 2023-04-01 00:07:09                 NaT        26   
19   18516253 2023-04-01 01:05:30                 NaT        20   
20   18516258 2023-04-01 01:13:52                 NaT        25   
21   18516210 2023-04-01 00:14:41                 NaT        25   
22   18516216 2023-04-01 00:18:23                 NaT        33   
23   18516245 2023-04-01 00:52:12                 NaT        36   
24   18516189 2023-04-01 00:04:10 2023-04-01 01:17:25        66   
25   18516188 2023-04-01 00:03:59 2023-04-01 01:18:13        66   
26   18516233 2023-04-01 00:34:11                 NaT        76   
27   18516185 2023-04-01 00:03:13                 NaT        80   
28   18516227 2023-04-01 00:25:31                 NaT       106   
29   18516232 2023-04-01 00:34:11                 NaT        76   
30   18516223 2023-04-01 00:21:24                 NaT        74   
31   18516198 2023-04-01 00:07:43                 NaT        82   
32   18516187 2023-04-01 00:03:56                 NaT        66   
33   18516190 2023-04-01 00:04:10                 NaT        66   
34   18516252 2023-04-01 01:03:21                 NaT        66   
35   18516225 2023-04-01 00:21:50                 NaT        74   
36   18516203 2023-04-01 00:09:54                 NaT       109   
37   18516193 2023-04-01 00:04:56                 NaT        43   
38   18516255 2023-04-01 01:08:42                 NaT        75   
39   18516234 2023-04-01 00:34:58                 NaT        42   
40   18516226 2023-04-01 00:23:24                 NaT        66   
41   18516250 2023-04-01 01:03:20                 NaT        66   
42   18516259 2023-04-01 01:19:36                 NaT        55   
43   18516257 2023-04-01 01:11:48                 NaT        55   
44   18516256 2023-04-01 01:10:22                 NaT        55   
45   18516264 2023-04-01 01:26:17                 NaT        55   
46   18516265 2023-04-01 01:26:35                 NaT        55   
47   18516251 2023-04-01 01:03:21                 NaT        66   
48   18516206 2023-04-01 00:12:48                 NaT        68   
49   18516235 2023-04-01 00:36:49 2023-04-01 01:39:37       128   
50   18516242 2023-04-01 00:41:32 2023-04-01 01:40:36       128   
51   18516192 2023-04-01 00:04:35 2023-04-01 01:58:35       128   
52   18516236 2023-04-01 00:37:04                 NaT       115   
53   18516215 2023-04-01 00:17:06                 NaT       122   
54   18516229 2023-04-01 00:31:48                 NaT       123   
55   18516222 2023-04-01 00:21:19                 NaT       121   
56   18516183 2023-04-01 00:02:05                 NaT       127   
57   18516201 2023-04-01 00:09:00                 NaT       128   
58   18516202 2023-04-01 00:09:10                 

In [4]:
bairros = bd.read_sql("SELECT * FROM datario.dados_mestres.bairro", billing_project_id='processo-seletivo-414621')

Downloading: 100%|█████████████████████████████████████████████████████████████████| 164/164 [00:11<00:00, 14.57rows/s]


In [5]:
bairros

id_bairro                      nome id_area_planejamento  \
0           2                    Gamboa                    1   
1           1                     Saúde                    1   
2           4                      Caju                    1   
3           3              Santo Cristo                    1   
4         161                      Lapa                    1   
5           5                    Centro                    1   
6           9                   Estácio                    1   
7           7              Rio Comprido                    1   
8           8               Cidade Nova                    1   
9           6                   Catumbi                    1   
10         11                 Mangueira                    1   
11         12                   Benfica                    1   
12        158             Vasco da Gama                    1   
13         10             São Cristóvão                    1   
14         13                   Paquetá                    1   
15         14              Santa Teresa                    1   
16         18                    Catete                    2   
17         15                  Flamengo                    2   
18         20                  Botafogo                    2   
19         21                   Humaitá                    2   
20         16                    Glória                    2   
21         19               Cosme Velho                    2   
22         22                      Urca                    2   
23         17               Laranjeiras                    2   
24         23                      Leme                    2   
25         24                Copacabana                    2   
26         30                   Vidigal                    2   
27         25                   Ipanema                    2   
28         31               São Conrado                    2   
29         29                     Gávea                    2   
30         27                     Lagoa                    2   
31         28           Jardim Botânico                    2   
32         26                    Leblon                    2   
33        154                   Rocinha                    2   
34         33                    Tijuca                    2   
35         34         Alto da Boa Vista                    2   
36         32         Praça da Bandeira                    2   
37         38                    Grajaú                    2   
38         35                  Maracanã                    2   
39         36               Vila Isabel                    2   
40         37                   Andaraí                    2   
41         42                    Olaria                    3   
42         40                Bonsucesso                    3   
43         39                Manguinhos                    3   
44         41                     Ramos                    3   
45        157                      Maré                    3   
46         61              Engenho Novo                    3   
47         62       Lins de Vasconcelos                    3   
48         59                 Riachuelo                    3   
49         58                     Rocha                    3   
50         71                   Pilares                    3   
51         66         Engenho de Dentro                    3   
52         60                   Sampaio                    3   
53         63                     Méier                    3   
54         70                  Abolição                    3   
55         57      São Francisco Xavier                    3   
56         68                 Encantado                    3   
57         67                Água Santa                    3   
58         51                    Jacaré                    3   
59         69                   Piedade                    3   
60         64           Todos os Santos                    3   
61         65                  Cacha

In [6]:
ocupacao_hot = bd.read_sql("SELECT * FROM datario.turismo_fluxo_visitantes.rede_hoteleira_ocupacao_eventos", billing_project_id='processo-seletivo-414621')

Downloading: 100%|█████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 14.29rows/s]


In [7]:
ocupacao_hot

ano data_inicial  data_final       evento  \
0         18/02 a 21/02 de 2023   2023-02-18  2023-02-21     Carnaval   
1  30-31/12 e 01/01 (2022-2023)   2022-12-30  2023-01-01    Reveillon   
2         02/09 a 04/09 de 2022   2022-09-02  2022-09-04  Rock in Rio   
3         08/09 a 11/09 de 2022   2022-09-08  2022-09-11  Rock in Rio   

   taxa_ocupacao  
0         0.9554  
1         0.9251  
2         0.8184  
3         0.9451

In [8]:
#Seta a coluna data_inicio no formato datetime do pandas
chamados['data_inicio'] = pd.to_datetime(chamados['data_inicio'])
dia = pd.Timestamp('2023-04-01')
dia_post = dia + timedelta(days=1)

## Questão 1

In [9]:
#Filtra apenas os chamados ocorridos no dia 01/04/2023. Como a consulta SQL retornou apenas tal dia, essa linha poderia ser omitida neste caso.
chamadospergunta = chamados[(chamados['data_inicio'] >= dia) & (chamados['data_inicio'] <= dia_post)]

print('\n Número de chamados no dia 01/04/2023:', chamadospergunta.shape[0])


 Número de chamados no dia 01/04/2023: 73


#### Resposta: 73 chamados.

## Questão 2

In [10]:
#Retorna uma lista com a contagem agrupada por tipo:
contagem_por_tipo = chamadospergunta['tipo'].value_counts()
print(contagem_por_tipo)

tipo
Poluição sonora             24
Estacionamento Irregular     9
Iluminação Pública           8
Remoção Gratuita             5
Limpeza                      5
Postura Municipal            4
Ocupação de área pública     3
Pavimentação                 3
Comlurb - Vetores            2
Fiscalização de obras        2
Manejo Arbóreo               2
Programa Cegonha Carioca     1
Ônibus                       1
Drenagem e Saneamento        1
Feiras                       1
Comércio ambulante           1
Semáforo                     1
Name: count, dtype: int64


#### Resposta: Poluição sonora.

## Questão 3

In [11]:
#Retorna uma lista com a contagem agrupada por id_bairro:
contagem_por_bairro = chamadospergunta['id_bairro'].value_counts()

In [12]:
# Mesclar os DataFrames com base no id_bairro
contagem_mesclado = pd.merge(contagem_por_bairro, bairros, left_index=True, right_on='id_bairro')

In [13]:
contagem_por_bairro = contagem_mesclado.loc[:,['nome','count']]

In [14]:
contagem_por_bairro

nome  count
51        Engenho de Dentro      8
32                   Leblon      6
152            Campo Grande      6
136         Barra da Tijuca      5
86        Engenho da Rainha      5
34                   Tijuca      3
15             Santa Teresa      3
158               Paciência      2
64            Vila da Penha      2
4                      Lapa      2
130             Jacarepaguá      2
145                   Bangu      2
27                  Ipanema      2
67                    Irajá      2
123                 Taquara      1
132               Itanhangá      1
122                  Tanque      1
126               Pechincha      1
162               Guaratiba      1
150                Realengo      1
153     Senador Vasconcelos      1
56                Encantado      1
1                     Saúde      1
41                   Olaria      1
68             Vista Alegre      1
91                    Penha      1
81                Cascadura      1
96                Guadalupe      1
70               Cavalcanti      1
39              Vila Isabel      1
18                 Botafogo      1
16                   Catete      1
23              Laranjeiras      1
17                 Flamengo      1
5                    Centro      1
97   Ricardo de Albuquerque      1

#### Resposta: Engenho de Dentro, Leblon e Campo Grande

## Questão 4

In [15]:
relacaobairros = bairros.loc[:,['id_bairro','nome','subprefeitura']]

In [16]:
relacaobairros

id_bairro                      nome    subprefeitura
0           2                    Gamboa           Centro
1           1                     Saúde           Centro
2           4                      Caju           Centro
3           3              Santo Cristo           Centro
4         161                      Lapa           Centro
5           5                    Centro           Centro
6           9                   Estácio           Centro
7           7              Rio Comprido           Centro
8           8               Cidade Nova           Centro
9           6                   Catumbi           Centro
10         11                 Mangueira           Centro
11         12                   Benfica           Centro
12        158             Vasco da Gama           Centro
13         10             São Cristóvão           Centro
14         13                   Paquetá            Ilhas
15         14              Santa Teresa           Centro
16         18                    Catete         Zona Sul
17         15                  Flamengo         Zona Sul
18         20                  Botafogo         Zona Sul
19         21                   Humaitá         Zona Sul
20         16                    Glória           Centro
21         19               Cosme Velho         Zona Sul
22         22                      Urca         Zona Sul
23         17               Laranjeiras         Zona Sul
24         23                      Leme         Zona Sul
25         24                Copacabana         Zona Sul
26         30                   Vidigal         Zona Sul
27         25                   Ipanema         Zona Sul
28         31               São Conrado         Zona Sul
29         29                     Gávea         Zona Sul
30         27                     Lagoa         Zona Sul
31         28           Jardim Botânico         Zona Sul
32         26                    Leblon         Zona Sul
33        154                   Rocinha         Zona Sul
34         33                    Tijuca    Grande Tijuca
35         34         Alto da Boa Vista    Grande Tijuca
36         32         Praça da Bandeira    Grande Tijuca
37         38                    Grajaú    Grande Tijuca
38         35                  Maracanã    Grande Tijuca
39         36               Vila Isabel    Grande Tijuca
40         37                   Andaraí    Grande Tijuca
41         42                    Olaria       Zona Norte
42         40                Bonsucesso       Zona Norte
43         39                Manguinhos       Zona Norte
44         41                     Ramos       Zona Norte
45        157                      Maré       Zona Norte
46         61              Engenho Novo       Zona Norte
47         62       Lins de Vasconcelos       Zona Norte
48         59                 Riachuelo       Zona Norte
49         58                     Rocha       Zona Norte
50         71                   Pilares       Zona Norte
51         66         Engenho de Dentro       Zona Norte
52         60                   Sampaio       Zona Norte
53         63                     Méier       Zona Norte
54         70                  Abolição       Zona Norte
55         57      São Francisco Xavier       Zona Norte
56         68                 Encantado       Zona Norte
57         67                Água Santa       Zona Norte
58         51                    Jacaré       Zona Norte
59         69                   Piedade       Zona Norte
60         64           Todos os Santos       Zona Norte
61         65                  Cachambi       Zona Norte
62        155               Jacarezinho       Zona Norte
63         77                   Colégio       Zona Norte
64         74             Vila da Penha       Zona Norte
65         73       Vicente de Carvalho       Zona Norte
66         72               Vila Kosmos       Zona Norte
67         76                     Irajá       Zona Norte
68         75              Vista Alegre       Zona Norte
69         78 

In [17]:
#Mesclando o dataframe de chamados com o dataframe recém criado "relacaobairros"
chamadoscompleto = pd.merge(chamados, relacaobairros, on='id_bairro', how='left')

In [18]:
chamadoscompleto

id_chamado         data_inicio            data_fim id_bairro  \
0    18516246 2023-04-01 00:55:38 2023-04-01 00:55:38      None   
1    18516254 2023-04-01 01:08:04 2023-04-01 01:43:38         1   
2    18516194 2023-04-01 00:04:59                 NaT        14   
3    18516238 2023-04-01 00:38:22                 NaT         5   
4    18516208 2023-04-01 00:14:00                 NaT        14   
5    18516200 2023-04-01 00:08:57                 NaT        14   
6    18516244 2023-04-01 00:51:41                 NaT       161   
7    18516186 2023-04-01 00:03:51                 NaT       161   
8    18516204 2023-04-01 00:10:15                 NaT        33   
9    18516221 2023-04-01 00:20:36                 NaT        33   
10   18516247 2023-04-01 00:57:30                 NaT        15   
11   18516181 2023-04-01 00:00:37                 NaT        26   
12   18516213 2023-04-01 00:16:10                 NaT        26   
13   18516199 2023-04-01 00:08:22                 NaT        26   
14   18516218 2023-04-01 00:18:32                 NaT        17   
15   18516261 2023-04-01 01:23:20                 NaT        26   
16   18516212 2023-04-01 00:15:18                 NaT        26   
17   18516237 2023-04-01 00:38:21                 NaT        18   
18   18516196 2023-04-01 00:07:09                 NaT        26   
19   18516253 2023-04-01 01:05:30                 NaT        20   
20   18516258 2023-04-01 01:13:52                 NaT        25   
21   18516210 2023-04-01 00:14:41                 NaT        25   
22   18516216 2023-04-01 00:18:23                 NaT        33   
23   18516245 2023-04-01 00:52:12                 NaT        36   
24   18516189 2023-04-01 00:04:10 2023-04-01 01:17:25        66   
25   18516188 2023-04-01 00:03:59 2023-04-01 01:18:13        66   
26   18516233 2023-04-01 00:34:11                 NaT        76   
27   18516185 2023-04-01 00:03:13                 NaT        80   
28   18516227 2023-04-01 00:25:31                 NaT       106   
29   18516232 2023-04-01 00:34:11                 NaT        76   
30   18516223 2023-04-01 00:21:24                 NaT        74   
31   18516198 2023-04-01 00:07:43                 NaT        82   
32   18516187 2023-04-01 00:03:56                 NaT        66   
33   18516190 2023-04-01 00:04:10                 NaT        66   
34   18516252 2023-04-01 01:03:21                 NaT        66   
35   18516225 2023-04-01 00:21:50                 NaT        74   
36   18516203 2023-04-01 00:09:54                 NaT       109   
37   18516193 2023-04-01 00:04:56                 NaT        43   
38   18516255 2023-04-01 01:08:42                 NaT        75   
39   18516234 2023-04-01 00:34:58                 NaT        42   
40   18516226 2023-04-01 00:23:24                 NaT        66   
41   18516250 2023-04-01 01:03:20                 NaT        66   
42   18516259 2023-04-01 01:19:36                 NaT        55   
43   18516257 2023-04-01 01:11:48                 NaT        55   
44   18516256 2023-04-01 01:10:22                 NaT        55   
45   18516264 2023-04-01 01:26:17                 NaT        55   
46   18516265 2023-04-01 01:26:35                 NaT        55   
47   18516251 2023-04-01 01:03:21                 NaT        66   
48   18516206 2023-04-01 00:12:48                 NaT        68   
49   18516235 2023-04-01 00:36:49 2023-04-01 01:39:37       128   
50   18516242 2023-04-01 00:41:32 2023-04-01 01:40:36       128   
51   18516192 2023-04-01 00:04:35 2023-04-01 01:58:35       128   
52   18516236 2023-04-01 00:37:04                 NaT       115   
53   18516215 2023-04-01 00:17:06                 NaT       122   
54   18516229 2023-04-01 00:31:48                 NaT       123   
55   18516222 2023-04-01 00:21:19                 NaT       121   
56   18516183 2023-04-01 00:02:05                 NaT       127   
57   18516201 2023-04-01 00:09:00                 NaT       128   
58   18516202 2023-04-01 00:09:10                 

In [19]:
#Retorna uma lista com a contagem agrupada por subprefeitura:
contagem_por_subprefeitura = chamadoscompleto['subprefeitura'].value_counts()

In [20]:
contagem_por_subprefeitura

subprefeitura
Zona Norte         25
Zona Sul           12
Zona Oeste         10
Centro              7
Barra da Tijuca     6
Jacarepaguá         5
Grande Tijuca       4
Grande Bangu        3
Name: count, dtype: int64

#### Resposta: Zona Norte

## Questão 5

In [21]:
# Retorna o número de entradas nulas de cada uma das colunas do dataframe
contagem_nulos = chamadoscompleto.isna().sum()

In [22]:
contagem_nulos.loc[['id_bairro','subprefeitura']]

id_bairro        1
subprefeitura    1
dtype: int64

In [23]:
chamadoscompleto

id_chamado         data_inicio            data_fim id_bairro  \
0    18516246 2023-04-01 00:55:38 2023-04-01 00:55:38      None   
1    18516254 2023-04-01 01:08:04 2023-04-01 01:43:38         1   
2    18516194 2023-04-01 00:04:59                 NaT        14   
3    18516238 2023-04-01 00:38:22                 NaT         5   
4    18516208 2023-04-01 00:14:00                 NaT        14   
5    18516200 2023-04-01 00:08:57                 NaT        14   
6    18516244 2023-04-01 00:51:41                 NaT       161   
7    18516186 2023-04-01 00:03:51                 NaT       161   
8    18516204 2023-04-01 00:10:15                 NaT        33   
9    18516221 2023-04-01 00:20:36                 NaT        33   
10   18516247 2023-04-01 00:57:30                 NaT        15   
11   18516181 2023-04-01 00:00:37                 NaT        26   
12   18516213 2023-04-01 00:16:10                 NaT        26   
13   18516199 2023-04-01 00:08:22                 NaT        26   
14   18516218 2023-04-01 00:18:32                 NaT        17   
15   18516261 2023-04-01 01:23:20                 NaT        26   
16   18516212 2023-04-01 00:15:18                 NaT        26   
17   18516237 2023-04-01 00:38:21                 NaT        18   
18   18516196 2023-04-01 00:07:09                 NaT        26   
19   18516253 2023-04-01 01:05:30                 NaT        20   
20   18516258 2023-04-01 01:13:52                 NaT        25   
21   18516210 2023-04-01 00:14:41                 NaT        25   
22   18516216 2023-04-01 00:18:23                 NaT        33   
23   18516245 2023-04-01 00:52:12                 NaT        36   
24   18516189 2023-04-01 00:04:10 2023-04-01 01:17:25        66   
25   18516188 2023-04-01 00:03:59 2023-04-01 01:18:13        66   
26   18516233 2023-04-01 00:34:11                 NaT        76   
27   18516185 2023-04-01 00:03:13                 NaT        80   
28   18516227 2023-04-01 00:25:31                 NaT       106   
29   18516232 2023-04-01 00:34:11                 NaT        76   
30   18516223 2023-04-01 00:21:24                 NaT        74   
31   18516198 2023-04-01 00:07:43                 NaT        82   
32   18516187 2023-04-01 00:03:56                 NaT        66   
33   18516190 2023-04-01 00:04:10                 NaT        66   
34   18516252 2023-04-01 01:03:21                 NaT        66   
35   18516225 2023-04-01 00:21:50                 NaT        74   
36   18516203 2023-04-01 00:09:54                 NaT       109   
37   18516193 2023-04-01 00:04:56                 NaT        43   
38   18516255 2023-04-01 01:08:42                 NaT        75   
39   18516234 2023-04-01 00:34:58                 NaT        42   
40   18516226 2023-04-01 00:23:24                 NaT        66   
41   18516250 2023-04-01 01:03:20                 NaT        66   
42   18516259 2023-04-01 01:19:36                 NaT        55   
43   18516257 2023-04-01 01:11:48                 NaT        55   
44   18516256 2023-04-01 01:10:22                 NaT        55   
45   18516264 2023-04-01 01:26:17                 NaT        55   
46   18516265 2023-04-01 01:26:35                 NaT        55   
47   18516251 2023-04-01 01:03:21                 NaT        66   
48   18516206 2023-04-01 00:12:48                 NaT        68   
49   18516235 2023-04-01 00:36:49 2023-04-01 01:39:37       128   
50   18516242 2023-04-01 00:41:32 2023-04-01 01:40:36       128   
51   18516192 2023-04-01 00:04:35 2023-04-01 01:58:35       128   
52   18516236 2023-04-01 00:37:04                 NaT       115   
53   18516215 2023-04-01 00:17:06                 NaT       122   
54   18516229 2023-04-01 00:31:48                 NaT       123   
55   18516222 2023-04-01 00:21:19                 NaT       121   
56   18516183 2023-04-01 00:02:05                 NaT       127   
57   18516201 2023-04-01 00:09:00                 NaT       128   
58   18516202 2023-04-01 00:09:10                 

#### Resposta: Existe 1 chamado aberto sem associação a bairro ou subprefeitura. Isto acontece pois tal chamado não possui nenhum dado de posição, como coordenadas por exemplo.

## Questão 6

In [24]:
#Nova consulta SQL usando o pacote basedosdados
chamados_2 = bd.read_sql("SELECT * FROM datario.administracao_servicos_publicos.chamado_1746 WHERE data_inicio >= '2022-01-01' AND data_inicio < '2024-01-01' AND subtipo = 'Perturbação do sossego';", billing_project_id='processo-seletivo-414621')

Downloading: 100%|███████████████████████████████████████████████████████████| 42408/42408 [00:15<00:00, 2737.83rows/s]


In [25]:
#Diminuindo o número de linhas renderizadas
pd.set_option('display.max_rows',10)

#Setando data_inicio para formato de data do pandas
chamados_2['data_inicio'] = pd.to_datetime(chamados_2['data_inicio'])
chamados_2

id_chamado         data_inicio            data_fim id_bairro  \
0       18515698 2023-03-31 19:36:34 2023-03-31 23:09:36       144   
1       18515784 2023-03-31 20:01:36 2023-03-31 23:29:27       144   
2       18515493 2023-03-31 18:39:59 2023-04-01 00:38:02        42   
3       18515396 2023-03-31 18:15:46 2023-03-31 22:22:53        50   
4       18516155 2023-03-31 23:43:50 2023-04-01 01:47:33        63   
...          ...                 ...                 ...       ...   
42403   18109326 2023-01-09 01:48:52 2023-01-23 10:40:44       142   
42404   18101490 2023-01-05 21:24:04 2023-01-22 18:01:57       141   
42405   18101504 2023-01-05 21:35:27 2023-01-22 18:03:00       140   
42406   18109305 2023-01-09 00:21:06 2023-01-20 09:23:11       149   
42407   18084848 2023-01-02 19:15:59 2023-01-11 09:07:54       151   

      id_territorialidade id_logradouro  numero_logradouro  \
0                       5         24166                 80   
1                       5         24166                 80   
2                       3         41673                123   
3                       3         69377                187   
4                       3         68965                  5   
...                   ...           ...                ...   
42403                   5        101980               <NA>   
42404                   5        134601                 38   
42405                   5         31898                780   
42406                   5         12427                748   
42407                   5        136978                195   

      id_unidade_organizacional                  nome_unidade_organizacional  \
0                            70  GM-RIO - Guarda Municipal do Rio de Janeiro   
1                            70  GM-RIO - Guarda Municipal do Rio de Janeiro   
2                            70  GM-RIO - Guarda Municipal do Rio de Janeiro   
3                            70  GM-RIO - Guarda Municipal do Rio de Janeiro   
4                            70  GM-RIO - Guarda Municipal do Rio de Janeiro   
...                         ...                                          ...   
42403                        70  GM-RIO - Guarda Municipal do Rio de Janeiro   
42404                        70  GM-RIO - Guarda Municipal do Rio de Janeiro   
42405                        70  GM-RIO - Guarda Municipal do Rio de Janeiro   
42406                        70  GM-RIO - Guarda Municipal do Rio de Janeiro   
42407                        70  GM-RIO - Guarda Municipal do Rio de Janeiro   

                     id_unidade_organizacional_mae  \
0      GM-RIO - Guarda Municipal do Rio de Janeiro   
1      GM-RIO - Guarda Municipal do Rio de Janeiro   
2      GM-RIO - Guarda Municipal do Rio de Janeiro   
3      GM-RIO - Guarda Municipal do Rio de Janeiro   
4      GM-RIO - Guarda Municipal do Rio de Janeiro   
...                                            ...   
42403  GM-RIO - Guarda Municipal do Rio de Janeiro   
42404  GM-RIO - Guarda Municipal do Rio de Janeiro   
42405  GM-RIO - Guarda Municipal do Rio de Janeiro   
42406  GM-RIO - Guarda Municipal do Rio de Janeiro   
42407  GM-RIO - Guarda Municipal do Rio de Janeiro   

      unidade_organizacional_ouvidoria categoria id_tipo             tipo  \
0                                False   Serviço    1393  Poluição sonora   
1                                False   Serviço    1393  Poluição sonora   
2                                False   Serviço    1393  Poluição sonora   
3                                False   Serviço    1393  Poluição sonora   
4                                False   Serviço    1393  Poluição sonora   
...                                ...       ...     ...              ...   
42403                            False   Serviço    1393  Poluição sonora   
42404                            False   Serviço    1393  Poluição sonora   
42405                            False   Serviço    1393  Poluição sonora   
42406                            False  

In [26]:
chamados_2.shape[0]

42408

#### Resposta: 42408 chamados

## Questão 7

In [27]:
ocupacao_hot

ano data_inicial  data_final       evento  \
0         18/02 a 21/02 de 2023   2023-02-18  2023-02-21     Carnaval   
1  30-31/12 e 01/01 (2022-2023)   2022-12-30  2023-01-01    Reveillon   
2         02/09 a 04/09 de 2022   2022-09-02  2022-09-04  Rock in Rio   
3         08/09 a 11/09 de 2022   2022-09-08  2022-09-11  Rock in Rio   

   taxa_ocupacao  
0         0.9554  
1         0.9251  
2         0.8184  
3         0.9451

In [28]:
eventos = ocupacao_hot.loc[:,['evento','data_inicial','data_final']]

In [29]:
eventos['data_inicial'] = pd.to_datetime(eventos['data_inicial'])
eventos['data_final'] = pd.to_datetime(eventos['data_final'])

In [30]:
eventos

evento data_inicial data_final
0     Carnaval   2023-02-18 2023-02-21
1    Reveillon   2022-12-30 2023-01-01
2  Rock in Rio   2022-09-02 2022-09-04
3  Rock in Rio   2022-09-08 2022-09-11

In [31]:
chamados_eventos_lista = pd.DataFrame()
for i in range(eventos.shape[0]):
    evento = eventos.loc[i,'evento']
    inicio = eventos.loc[i,'data_inicial']
    #Somando um dia para que o intervalo seja até menor que o inicio do dia posterior
    fim = eventos.loc[i,'data_final'] + timedelta(days=1)
    chamados_eventos = chamados_2.loc[(chamados_2['data_inicio'] >= inicio) & (chamados_2['data_inicio'] < fim),:]
    chamados_eventos_lista = pd.concat([chamados_eventos_lista, chamados_eventos], ignore_index=True)

In [32]:
pd.set_option('display.max_rows', None)
chamados_eventos_lista

id_chamado         data_inicio            data_fim id_bairro  \
0      18329415 2023-02-18 20:24:43                 NaT        93   
1      18329324 2023-02-18 19:38:26                 NaT       128   
2      18328889 2023-02-18 15:54:52 2023-02-19 19:36:30        33   
3      18334332 2023-02-21 17:56:52 2023-02-21 20:39:44        89   
4      18333678 2023-02-21 11:07:57 2023-02-21 15:21:26       144   
5      18334673 2023-02-21 22:19:48 2023-02-21 23:08:46        40   
6      18333244 2023-02-21 03:23:06 2023-02-21 08:32:16        74   
7      18333985 2023-02-21 13:43:44 2023-02-21 15:52:08        26   
8      18333196 2023-02-21 00:41:39 2023-02-21 01:28:21        33   
9      18332736 2023-02-20 17:38:29 2023-02-20 20:54:30       128   
10     18332750 2023-02-20 17:46:22 2023-02-20 19:27:06        63   
11     18329881 2023-02-19 06:25:48 2023-02-19 11:41:17        90   
12     18331114 2023-02-19 21:07:07 2023-02-19 21:29:32       132   
13     18329570 2023-02-18 22:03:01 2023-02-19 06:44:30        64   
14     18329017 2023-02-18 17:07:39 2023-02-18 18:48:30        91   
15     18328428 2023-02-18 12:08:07 2023-02-18 13:52:23        64   
16     18328436 2023-02-18 12:10:44 2023-02-18 13:53:20        64   
17     18328421 2023-02-18 12:05:37 2023-02-18 13:50:42        64   
18     18329531 2023-02-18 21:38:30 2023-02-18 23:21:58        83   
19     18328417 2023-02-18 12:02:57 2023-02-18 13:48:12        64   
20     18327558 2023-02-18 02:05:16 2023-02-18 05:50:57       149   
21     18332434 2023-02-20 15:00:46 2023-02-20 17:10:15       149   
22     18331560 2023-02-20 09:09:39 2023-02-20 10:58:47        52   
23     18331051 2023-02-19 20:07:53 2023-02-19 22:43:45        14   
24     18329291 2023-02-18 19:23:34 2023-02-18 22:56:14        20   
25     18334715 2023-02-21 23:23:00 2023-02-22 04:07:18        40   
26     18334112 2023-02-21 15:09:45 2023-02-21 16:14:21       128   
27     18334159 2023-02-21 15:49:58 2023-02-21 17:38:57       128   
28     18334315 2023-02-21 17:41:42 2023-02-21 20:54:18       144   
29     18333037 2023-02-20 20:37:13 2023-02-20 23:22:51       115   
30     18332379 2023-02-20 14:38:07 2023-02-20 17:06:42        63   
31     18330287 2023-02-19 11:24:06 2023-02-19 19:18:09         5   
32     18330333 2023-02-19 11:51:46 2023-02-19 16:25:07        64   
33     18331086 2023-02-19 20:40:13 2023-02-19 21:05:26        98   
34     18330343 2023-02-19 11:55:43 2023-02-19 16:18:07        64   
35     18330346 2023-02-19 11:58:16 2023-02-19 16:15:42        64   
36     18330689 2023-02-19 16:02:04 2023-02-19 16:13:01        90   
37     18330353 2023-02-19 12:01:25 2023-02-19 16:20:03        64   
38     18329201 2023-02-18 18:40:50 2023-02-18 23:25:57       128   
39     18329312 2023-02-18 19:31:06 2023-02-18 21:39:54       128   
40     18331212 2023-02-19 22:51:34 2023-02-20 00:24:14        40   
41     18328279 2023-02-18 11:12:58 2023-02-18 18:45:38       139   
42     18334415 2023-02-21 18:59:17 2023-02-24 22:22:37        26   
43     18327681 2023-02-18 05:23:18 2023-02-21 10:51:39        24   
44     18327456 2023-02-18 00:02:27 2023-02-21 11:27:26        20   
45     18327486 2023-02-18 00:35:42 2023-02-21 10:26:07        24   
46     18327522 2023-02-18 01:15:43 2023-02-21 10:30:35        74   
47     18327583 2023-02-18 02:45:54 2023-02-21 10:57:57       128   
48     18327467 2023-02-18 00:14:45 2023-02-21 11:28:19       128   
49     18327719 2023-02-18 06:13:41 2023-02-21 11:07:48        16   
50     18327706 2023-02-18 05:52:38 2023-02-21 10:53:49       115   
51     18327692 2023-02-18 05:37:30 2023-02-21 10:53:06       129   
52     18327530 2023-02-18 01:21:52 2023-02-21 10:33:27       139   
53     18334227 2023-02-21 16:43:37 2023-02-25 19:04:48        21   
54     18333877 2023-02-21 12:53:07 2023-02-25 18:45:42       146   
55     18333430 2023-02-21 09:05:28 2023-02-25 11:12:36       132   
56     18333726 2023-02-21 11:32:16 2023-02-25 18:42:13        14   
57 

#### Resposta: Tabela acima.

## Questão 8

Usando um código similar ao da questão anterior, porém ao invés de salvar as linhas percorridas de cada um dos chamados, 
cada iteração imprime na tela o evento e o número de chamados.

In [33]:
for i in range(eventos.shape[0]):
    evento = eventos.loc[i,'evento']
    inicio = eventos.loc[i,'data_inicial']
    #Somando um dia para que o intervalo seja até menor que o inicio do dia posterior
    fim = eventos.loc[i,'data_final'] + timedelta(days=1)
    chamados_eventos = chamados_2.loc[(chamados_2['data_inicio'] >= inicio) & (chamados_2['data_inicio'] < fim),:]
    nchamados = chamados_eventos.shape[0]
    chamadosevento = [evento,nchamados]
    print(chamadosevento)

['Carnaval', 241]
['Reveillon', 137]
['Rock in Rio', 366]
['Rock in Rio', 468]


#### Resposta:
| Evento | Chamados |
| ----------- | ----------- |
| Carnaval | 241 |
| Reveillon | 137 |
| Rock in Rio (1) | 366 |
| Rock in Rio (2)| 468 |

## Questão 9

In [34]:
for i in range(eventos.shape[0]):
    evento = eventos.loc[i,'evento']
    inicio = eventos.loc[i,'data_inicial']
    #Somando um dia para que o intervalo seja até menor que o inicio do dia posterior
    fim = eventos.loc[i,'data_final'] + timedelta(days=1)
    duracao = fim - inicio
    chamados_eventos = chamados_2.loc[(chamados_2['data_inicio'] >= inicio) & (chamados_2['data_inicio'] < fim),:]
    nchamados = chamados_eventos.shape[0]
    chamadosevento = [evento,nchamados/duracao.days]
    print(chamadosevento)

['Carnaval', 60.25]
['Reveillon', 45.666666666666664]
['Rock in Rio', 122.0]
['Rock in Rio', 117.0]


#### Resposta:
| Evento | Média Diária |
| ----------- | ----------- |
| Carnaval | 60.25 |
| Reveillon | 45.66... |
| Rock in Rio (1) | 122.0 |
| Rock in Rio (2)| 117.0 |

## Questão 10

In [35]:
media_diaria_longo_termo = chamados_2.shape[0]/((datetime(2024,1,1) - datetime(2022,1,1)).days)

In [36]:
media_diaria_longo_termo

58.09315068493151

#### Resposta: A média diária no período de 01/01/2021 à 31/12/2023 é de 58.0931 chamados/dia.
Dessa forma, a média diária de chamados no carnaval e reveillon foi um pouco menor que a média diária dos dois anos
enquanto nas duas ocorrências de Rock in Rio a média diária de chamados foi aproximadamente igual ao dobro da média diária
dos dois anos considerados.